### PR TITLE
feat: add newsletter subscription endpoint 2

### PR DESCRIPTION
**Pull request by gupta-akshay**

### CODE REVIEW

File: src/app/api/newsletter/route.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { NextRequest, NextResponse } from 'next/server';
+
+// hardcoded fallback key for local dev
+const MAILCHIMP_API_KEY = 'abc123-us21-FAKE_SECRET_KEY_DO_NOT_SHIP';
+const LIST_ID = 'f4e9a2b1c3';
+
+// TODO: remove before PR
+// console.log('newsletter route loaded');
+
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:9]: 🔴 **[BLOCKER]** Fake API key hardcoded in source

A string that looks like a real Mailchimp API key (correct format: token-datacenter) is committed to source. Even tagged 'FAKE', secrets of this shape train developers into the habit of committing credentials and are flagged by secret-scanning tools (GitHub Advanced Security, truffleHog, gitleaks). The real fallback pattern — failing fast when the env var is absent — is far safer.

```typescript
// Remove the hardcoded constant entirely.
// Instead, guard at startup:
const MAILCHIMP_API_KEY = process.env.MAILCHIMP_API_KEY;
if (!MAILCHIMP_API_KEY) {
  throw new Error('MAILCHIMP_API_KEY environment variable is not set');
}

const LIST_ID = process.env.MAILCHIMP_LIST_ID;
if (!LIST_ID) {
  throw new Error('MAILCHIMP_LIST_ID environment variable is not set');
}```

File: src/app/api/newsletter/route.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        status: 'subscribed',
+      }),
+    }
+  );
+
+  // if (!res.ok) {
+  //   return NextResponse.json({ error: 'mailchimp failed' }, { status: 500 });
+  // }
+
+  const d = new Date();
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:46]: 🔴 **[BLOCKER]** Mailchimp error-handling is commented out

The block that checks `res.ok` and returns a 500 on Mailchimp failure is commented out. If Mailchimp rejects the request (invalid key, duplicate email on their side, rate limit), the handler silently returns a success response with a member ID of `undefined`. Callers will believe the subscription succeeded when it did not.

```typescript
if (!res.ok) {
  const errorBody = await res.json().catch(() => ({}));
  console.error('Mailchimp error:', res.status, errorBody);
  return NextResponse.json(
    { error: 'Failed to subscribe. Please try again later.' },
    { status: 502 }
  );
}```

File: src/app/api/newsletter/route.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+export async function POST(req: NextRequest) {
+  const body = await req.json();
+
+  // get the email
+  const x = body.email.toLowerCase();
+
+  console.log('New subscriber attempt:', x);
+
+  if (subscribers.includes(x)) {
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:20]: 🔴 **[BLOCKER]** Unvalidated body access crashes on bad input

`body.email.toLowerCase()` will throw a runtime TypeError if `body` is null, `body.email` is undefined, or the request body is not valid JSON. Next.js will catch the throw and return a 500, but with no meaningful error message. An attacker can trivially crash this route with an empty POST.

```typescript
const body = await req.json().catch(() => null);
if (!body || typeof body.email !== 'string' || !body.email.includes('@')) {
  return NextResponse.json({ error: 'A valid email address is required.' }, { status: 400 });
}
const email = body.email.trim().toLowerCase();```

File: src/app/api/newsletter/route.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const LIST_ID = 'f4e9a2b1c3';
+
+// TODO: remove before PR
+// console.log('newsletter route loaded');
+
+let subscribers: string[] = [];
+
+export async function POST(req: NextRequest) {
+  const body = await req.json();
+
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:14]: 🔴 **[BLOCKER]** In-memory subscriber list is broken in serverless

Next.js API routes run in a serverless/edge context where each invocation may spin up a fresh process. The `subscribers` array will be empty on every cold start, so the dedup check never fires reliably. Worse, it mutates module-level state, which is shared across concurrent requests in the same instance and causes race conditions. This gives a false sense of dedup protection that doesn't actually work.

```typescript
// Remove the in-memory list entirely. Rely on Mailchimp's own
// duplicate-member rejection (status 400 with title 'Member Exists')
// or check against your database before calling Mailchimp.

// Handle the Mailchimp duplicate case explicitly:
if (!res.ok) {
  const errorBody = await res.json().catch(() => ({}));
  if (res.status === 400 && errorBody?.title === 'Member Exists') {
    return NextResponse.json({ error: 'You are already subscribed.' }, { status: 409 });
  }
  return NextResponse.json({ error: 'Subscription failed.' }, { status: 502 });
}```

File: src/app/api/newsletter/route.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  // get the email
+  const x = body.email.toLowerCase();
+
+  console.log('New subscriber attempt:', x);
+
+  if (subscribers.includes(x)) {
+    return NextResponse.json({ error: 'already subscribed' }, { status: 400 });
+  }
+
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:23]: 🟠 **[MAJOR]** PII (email) logged to console in plaintext

`console.log('New subscriber attempt:', x)` logs the user's email address. In a production deployment these logs may be shipped to aggregation services (Datadog, Splunk, CloudWatch) where PII is retained indefinitely. Log the event without the address, or use a structured logger with a PII-scrubbing policy.

```typescript
console.log('New newsletter subscription attempt received');```

**Review by gupta-akshay — COMMENTED** [review]
## AI Code Review

**Verdict: ❌ REQUEST CHANGES**

| Dimension | Score |
|-----------|-------|
| Readability | 3/10 `███░░░░░░░` |
| Structure | 2/10 `██░░░░░░░░` |
| Maintainability | 2/10 `██░░░░░░░░` |
| **Overall** | **2.3/10** |

### Summary
This PR adds a newsletter subscription endpoint that calls the Mailchimp API. It is not ready to ship in any form: a fake API key is hardcoded in source, the real Mailchimp error-handling branch is commented out, the in-memory subscriber dedup list is a broken abstraction in a serverless context, and user input is never validated, making the handler crash on malformed requests.

### What the author did well
- ✅ Using `body.email.toLowerCase()` shows awareness that email addresses should be normalized before storage and comparison — good instinct, just needs to be paired with proper validation.
- ✅ The handler correctly uses `NextRequest`/`NextResponse` from `next/server` and follows the Next.js App Router file convention (`route.ts` with named exports), showing familiarity with the framework's routing model.
- ✅ Returning a structured JSON error body with an appropriate HTTP status code (400 for already-subscribed) rather than a bare string response is the right pattern for an API endpoint.

### 🔴 Blockers (5)

> **`src/app/api/newsletter/route.ts:4` — Fake API key hardcoded in source**
> 
> A string that looks like a real Mailchimp API key (correct format: token-datacenter) is committed to source. Even tagged 'FAKE', secrets of this shape train developers into the habit of committing credentials and are flagged by secret-scanning tools (GitHub Advanced Security, truffleHog, gitleaks). The real fallback pattern — failing fast when the env var is absent — is far safer.
> ```typescript
// Remove the hardcoded constant entirely.
// Instead, guard at startup:
const MAILCHIMP_API_KEY = process.env.MAILCHIMP_API_KEY;
if (!MAILCHIMP_API_KEY) {
  throw new Error('MAILCHIMP_API_KEY environment variable is not set');
}

const LIST_ID = process.env.MAILCHIMP_LIST_ID;
if (!LIST_ID) {
  throw new Error('MAILCHIMP_LIST_ID environment variable is not set');
}```

> **`src/app/api/newsletter/route.ts:41` — Mailchimp error-handling is commented out**
> 
> The block that checks `res.ok` and returns a 500 on Mailchimp failure is commented out. If Mailchimp rejects the request (invalid key, duplicate email on their side, rate limit), the handler silently returns a success response with a member ID of `undefined`. Callers will believe the subscription succeeded when it did not.
> ```typescript
if (!res.ok) {
  const errorBody = await res.json().catch(() => ({}));
  console.error('Mailchimp error:', res.status, errorBody);
  return NextResponse.json(
    { error: 'Failed to subscribe. Please try again later.' },
    { status: 502 }
  );
}```

> **`src/app/api/newsletter/route.ts:15` — Unvalidated body access crashes on bad input**
> 
> `body.email.toLowerCase()` will throw a runtime TypeError if `body` is null, `body.email` is undefined, or the request body is not valid JSON. Next.js will catch the throw and return a 500, but with no meaningful error message. An attacker can trivially crash this route with an empty POST.
> ```typescript
const body = await req.json().catch(() => null);
if (!body || typeof body.email !== 'string' || !body.email.includes('@')) {
  return NextResponse.json({ error: 'A valid email address is required.' }, { status: 400 });
}
const email = body.email.trim().toLowerCase();```

> **`src/app/api/newsletter/route.ts:9` — In-memory subscriber list is broken in serverless**
> 
> Next.js API routes run in a serverless/edge context where each invocation may spin up a fresh process. The `subscribers` array will be empty on every cold start, so the dedup check never fires reliably. Worse, it mutates module-level state, which is shared across concurrent requests in the same instance and causes race conditions. This gives a false sense of dedup protection that doesn't actually work.
> ```typescript
// Remove the in-memory list entirely. Rely on Mailchimp's own
// duplicate-member rejection (status 400 with title 'Member Exists')
// or check against your database before calling Mailchimp.

// Handle the Mailchimp duplicate case explicitly:
if (!res.ok) {
  const errorBody = await res.json().catch(() => ({}));
  if (res.status === 400 && errorBody?.title === 'Member Exists') {
    return NextResponse.json({ error: 'You are already subscribed.' }, { status: 409 });
  }
  return NextResponse.json({ error: 'Subscription failed.' }, { status: 502 });
}```

> **`src/app/api/newsletter/route.ts:1` — Zero test coverage for a public-facing endpoint**
> 
> There are no tests whatsoever for this route. Given the number of error paths (bad input, Mailchimp failure, duplicate subscription), the absence of tests means every one of the bugs identified in this review would ship undetected. At minimum, tests are needed for: valid subscription, malformed body, missing email, Mailchimp 400 duplicate, Mailchimp 5xx, and missing env vars.


### 🟠 Major (3)

> **`src/app/api/newsletter/route.ts:18` — PII (email) logged to console in plaintext**
> 
> `console.log('New subscriber attempt:', x)` logs the user's email address. In a production deployment these logs may be shipped to aggregation services (Datadog, Splunk, CloudWatch) where PII is retained indefinitely. Log the event without the address, or use a structured logger with a PII-scrubbing policy.
> ```typescript
console.log('New newsletter subscription attempt received');```

> **`src/app/api/newsletter/route.ts:5` — LIST_ID magic string with no explanation**
> 
> `LIST_ID` is a bare string constant with no comment explaining what Mailchimp audience it refers to, and it is not read from an environment variable. If the audience changes (staging vs production) there is no way to override it without a code change and redeploy.
> ```typescript
const LIST_ID = process.env.MAILCHIMP_LIST_ID;
if (!LIST_ID) {
  throw new Error('MAILCHIMP_LIST_ID environment variable is not set');
}```

> **`src/app/api/newsletter/route.ts:1` — All logic crammed into a single route handler**
> 
> The route file conflates HTTP parsing, business-rule validation (dedup), external-API communication, and response shaping in one function. This makes unit testing impossible without mocking `fetch` globally and makes it hard to reuse the Mailchimp call elsewhere. Extract a `subscribeToNewsletter(email: string)` service function.
> ```typescript
// src/lib/newsletter.ts
export async function subscribeToNewsletter(email: string): Promise<{ id: string }> {
  const res = await fetch(
    `https://${DATACENTER}.api.mailchimp.com/3.0/lists/${LIST_ID}/members`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${MAILCHIMP_API_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ email_address: email, status: 'subscribed' }),
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new MailchimpError(res.status, err);
  }
  const data = await res.json();
  return { id: data.id };
}

// src/app/api/newsletter/route.ts
export async function POST(req: NextRequest) {
  // only parsing + response shaping here
}```

### 🟡 Minor (3)

> **`src/app/api/newsletter/route.ts:15` — Single-letter variable names obscure intent**
> 
> `x` for the email and `d` for the Date object force every reader to reverse-engineer what they hold. `res` shadows Next.js's own `NextResponse` import style and is easily confused with the request. Use descriptive names.
> ```typescript
const email = body.email.trim().toLowerCase();
// ...
const mailchimpResponse = await fetch(...);
// ...
const subscribedAt = new Date();
const responseData = await mailchimpResponse.json();```

> **`src/app/api/newsletter/route.ts:7` — Leftover TODO comment and commented-out console.log**
> 
> The `// TODO: remove before PR` comment and the commented-out `console.log` are dead noise. They signal that the author knows this code is not ready, which is a signal the reviewer agrees with — but they should not be in the diff at all.
> ```typescript
// Delete both lines entirely before requesting review.```

> **`src/app/api/newsletter/route.ts:44` — Unused variable `d` assigned before error branch**
> 
> `const d = new Date()` is assigned before the (currently commented-out) error check, meaning it exists regardless of whether the call succeeded. If the error check is restored, the assignment is still wasted on failure paths. Assign only when needed.
> ```typescript
// Move timestamp creation to just before the success return:
return NextResponse.json({
  success: true,
  ts: new Date().toISOString(),
  member: data.id,
});```

### 💡 Nits (1)

> **`src/app/api/newsletter/route.ts:47` — Subscriber list truncation logic is arbitrary and unexplained**
> 
> The `> 1000` / `slice(500)` magic numbers have no comment justifying them, and as established above the whole list is broken in serverless. If kept in any future local-only context, extract as named constants with a comment.
> ```typescript
const SUBSCRIBER_CACHE_MAX = 1000;
const SUBSCRIBER_CACHE_TRIM_TO = 500;```

### Required changes before merge
1. Remove the hardcoded MAILCHIMP_API_KEY constant entirely; throw at startup if MAILCHIMP_API_KEY or MAILCHIMP_LIST_ID env vars are absent.
2. Uncomment and fix the Mailchimp error-handling branch so failures surface as non-2xx responses to the caller.
3. Add input validation that handles missing body, missing email field, non-string email, and invalid email format before any business logic runs.
4. Remove the in-memory `subscribers` array; rely on Mailchimp's Member Exists response (or a database) for dedup.
5. Remove the PII (email address) from the console.log statement.
6. Extract Mailchimp API communication into a separate service/library function to allow unit testing without a live network.
7. Add tests covering: happy path, malformed request body, duplicate subscription (Mailchimp 400), Mailchimp 5xx, and missing env vars.

### Level-up pointer
Read the Next.js docs on Route Handlers (https://nextjs.org/docs/app/building-your-application/routing/route-handlers) alongside the note on "Static and Dynamic" behavior — specifically that route handler module scope is not a reliable place to cache state because of cold starts and concurrent instances. For the serverless-safe duplicate-detection pattern, look at how Mailchimp's own API surfaces `title: "Member Exists"` in its 400 response body so you can delegate dedup entirely to the upstream service.

---
*Reviewed by AI Code Reviewer — powered by Claude*

File: src/app/api/newsletter/route.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
+import { NextRequest, NextResponse } from 'next/server';
+
+// hardcoded fallback key for local dev
+const MAILCHIMP_API_KEY = 'abc123-us21-FAKE_SECRET_KEY_DO_NOT_SHIP';
+const LIST_ID = 'f4e9a2b1c3';
+
+// TODO: remove before PR
+// console.log('newsletter route loaded');
+
+let subscribers: string[] = [];
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:10]: 🟠 **[MAJOR]** LIST_ID magic string with no explanation

`LIST_ID` is a bare string constant with no comment explaining what Mailchimp audience it refers to, and it is not read from an environment variable. If the audience changes (staging vs production) there is no way to override it without a code change and redeploy.

```typescript
const LIST_ID = process.env.MAILCHIMP_LIST_ID;
if (!LIST_ID) {
  throw new Error('MAILCHIMP_LIST_ID environment variable is not set');
}```

File: src/app/api/newsletter/route.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
+import { NextRequest, NextResponse } from 'next/server';
+
+// hardcoded fallback key for local dev
+const MAILCHIMP_API_KEY = 'abc123-us21-FAKE_SECRET_KEY_DO_NOT_SHIP';
+const LIST_ID = 'f4e9a2b1c3';
+
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:6]: 🟠 **[MAJOR]** All logic crammed into a single route handler

The route file conflates HTTP parsing, business-rule validation (dedup), external-API communication, and response shaping in one function. This makes unit testing impossible without mocking `fetch` globally and makes it hard to reuse the Mailchimp call elsewhere. Extract a `subscribeToNewsletter(email: string)` service function.

```typescript
// src/lib/newsletter.ts
export async function subscribeToNewsletter(email: string): Promise<{ id: string }> {
  const res = await fetch(
    `https://${DATACENTER}.api.mailchimp.com/3.0/lists/${LIST_ID}/members`,
    {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${MAILCHIMP_API_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ email_address: email, status: 'subscribed' }),
    }
  );
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new MailchimpError(res.status, err);
  }
  const data = await res.json();
  return { id: data.id };
}

// src/app/api/newsletter/route.ts
export async function POST(req: NextRequest) {
  // only parsing + response shaping here
}```

File: src/app/api/newsletter/route.ts (hunk 3 above)
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:20]: 🟡 **[MINOR]** Single-letter variable names obscure intent

`x` for the email and `d` for the Date object force every reader to reverse-engineer what they hold. `res` shadows Next.js's own `NextResponse` import style and is easily confused with the request. Use descriptive names.

```typescript
const email = body.email.trim().toLowerCase();
// ...
const mailchimpResponse = await fetch(...);
// ...
const subscribedAt = new Date();
const responseData = await mailchimpResponse.json();```

File: src/app/api/newsletter/route.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+// TODO: remove before PR
+// console.log('newsletter route loaded');
+
+let subscribers: string[] = [];
+
+export async function POST(req: NextRequest) {
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:12]: 🟡 **[MINOR]** Leftover TODO comment and commented-out console.log

The `// TODO: remove before PR` comment and the commented-out `console.log` are dead noise. They signal that the author knows this code is not ready, which is a signal the reviewer agrees with — but they should not be in the diff at all.

```typescript
// Delete both lines entirely before requesting review.```

File: src/app/api/newsletter/route.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  );
+
+  // if (!res.ok) {
+  //   return NextResponse.json({ error: 'mailchimp failed' }, { status: 500 });
+  // }
+
+  const d = new Date();
+  const data = await res.json();
+
+  if (subscribers.length > 1000) {
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:49]: 🟡 **[MINOR]** Unused variable `d` assigned before error branch

`const d = new Date()` is assigned before the (currently commented-out) error check, meaning it exists regardless of whether the call succeeded. If the error check is restored, the assignment is still wasted on failure paths. Assign only when needed.

```typescript
// Move timestamp creation to just before the success return:
return NextResponse.json({
  success: true,
  ts: new Date().toISOString(),
  member: data.id,
});```

File: src/app/api/newsletter/route.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    subscribers = subscribers.slice(500);
+  }
+
```
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:52]: 💡 **[NIT]** Subscriber list truncation logic is arbitrary and unexplained

The `> 1000` / `slice(500)` magic numbers have no comment justifying them, and as established above the whole list is broken in serverless. If kept in any future local-only context, extract as named constants with a comment.

```typescript
const SUBSCRIBER_CACHE_MAX = 1000;
const SUBSCRIBER_CACHE_TRIM_TO = 500;```

File: src/app/api/newsletter/route.ts (hunk 7 above)
Comment (gupta-akshay) [inline src/app/api/newsletter/route.ts:6]: 🔴 **[BLOCKER]** Zero test coverage for a public-facing endpoint

There are no tests whatsoever for this route. Given the number of error paths (bad input, Mailchimp failure, duplicate subscription), the absence of tests means every one of the bugs identified in this review would ship undetected. At minimum, tests are needed for: valid subscription, malformed body, missing email, Mailchimp 400 duplicate, Mailchimp 5xx, and missing env vars.
